src/index.js

```diff
@@ -0,0 +1,87 @@
+'use strict';
+const assert = require('assert');
+const {Client} = require('remotely-signed-s3');
+
+class Artifact {
+  constructor({queue, client, clientOpts}) {
+    assert(typeof queue === 'object', 'Must provide a Queue client');
+    this.__queue = queue;
+
+    if (client && clientOpts) {
+      throw new Error('You can only provide a client or options for a client to be created with');
+    } else if (client) {
+      this.__client = client
+    } else if (clientOpts) {
+      this.__client = new Client(clientOpts);
+    } else {
+      this.__client = new Client();
+    }
+  }
+
+  /**
+   * Upload an artifact to the queue, optionally with compression
+   */
+  async put({taskId, runId, name, contentType, filename, compression, expires, forceMP, forceSP}) {
+    assert(typeof taskId === 'string', 'Must provide a taskId for upload');
+    assert(typeof runId === 'string', 'Must provide a runId for upload');
+    assert(typeof name === 'string', 'Must provide a name for upload');
+    assert(typeof contentType === 'string', 'Must provide a contentType for upload');
+    assert(typeof filename === 'string', 'Must provide a filename for upload');
+    if (compression) {
+      assert(typeof compression === 'string', 'Compression format must be string');
+      assert(compression === 'gzip', 'Only gzip compression is supported');
+    }
+    assert(typeof expires === 'object', 'Expires must be provided as a date object');
+    assert(expires.constructor.name === 'Date', 'Expires must be provided as a date object');
+
+    let upload = await this.__client.prepareUpload({filename, compression, forceMP, forceSP});
+
+    let requestParams = {
+      storageType: 'blob',
+      expires,
+      contentType,
+      contentSha256: upload.sha256,
+      contentLength: upload.size,
+    };
+
+    if (upload.contentEncoding && upload.contentEncoding !== 'identity') {
+      requestParams.transferSha256 = upload.transferSha256;
+      requestParams.transferLength = upload.transferSize;
+      requestParams.contentEncoding = upload.contentEncoding;
+    }
+
+    let requests = await this.__queue.createArtifact(taskId, runId, name, requestParams);
+    requests = requests.requests;
```

Comment:
Hmm, I'm not sure that's how it works.  JS is thinking you're trying to await on a `Promise` that's at the location `this.__queue.createArtifact(...).requests` .  You'd need parens to clarify to the JS engine which thing is the `Promise` vs. the properties you're looking up, and this line is already quite long.

```
jhford-work:~/taskcluster $ cat test.js
async function x() {
  return Promise.resolve({result: 1});
}

async function y() {
  let z = await x().result;
}

y().then(console.log, console.log);
jhford-work:~/taskcluster $ node --version
v8.2.1
jhford-work:~/taskcluster $ node test.js
undefined
jhford-work:~/taskcluster $ ./node_modules/.bin/babel-node test.js
undefined
```